### PR TITLE
feat(desk-tool): migrate desk-tool to `@sanity/ui`

### PR DESCRIPTION
**Pull request by mariuslundgard**
### Description

- Introducing new pane building block components to migrate the Desk Tool panes to Sanity UI. The new components are located in `packages/@sanity/desk-tool/src/components/pane`.
- All the Desk Tool panes have been migrated to use the new pane building blocks:
  - [x] `DocumentPane`
  - [x] `DocumentsListPane`
  - [x] `ErrorPane`
  - [x] `ListPane`
  - [x] `LoadingPane`
  - [x] `UnknownPane`
  - [x] `UserComponentPane`
- All code in `@sanity/desk-tool` has been converted to TypeScript (including fixing TS errors and linting).

### CODE REVIEW

File: packages/@sanity/desk-tool/src/lib/useUnique.ts
Comment (bjoerge) [inline packages/@sanity/desk-tool/src/lib/useUnique.ts:4]: could a better name for this be useDeepUnique to better reflect that it does a (potentially expensive) deep comparison on every render? I guess it's also worth adding a note about when *not* to use - I can easily imagine cases where rendering is cheap and using this will actually degrade performance.

File: packages/@sanity/desk-tool/src/lib/useShallowUnique.ts
Comment (bjoerge) [inline packages/@sanity/desk-tool/src/lib/useShallowUnique.ts:4]: would be great with a comment describing the purpose and an outline of what use case(s) this solves

File: packages/@sanity/desk-tool/src/components/pane/__workshop__/SplitPanes/DocumentPane/DocumentActions.tsx
Comment (bjoerge) [inline packages/@sanity/desk-tool/src/components/pane/__workshop__/SplitPanes/DocumentPane/DocumentActions.tsx:26]: put this in a const?

File: packages/@sanity/base/src/components/zOffsets/defaults.ts
Comment (bjoerge) [inline packages/@sanity/base/src/components/zOffsets/defaults.ts:26]: …so why is it still here? ;)

**Review by bjoerge — APPROVED** [review]
A lot of changes, so not able to go all that much in depth on everything, but looks good to me overall. Have only focused on reviewing code and not done any UI testing.

There seems to be a few commits that probably should be squashed / cleaned up before merging.

Left a few comments inline, mostly suggestions for improvements and nothing code wise that blocks merge, imo. Great work!

File: packages/@sanity/base/src/change-indicators/helpers/getOffsetsTo.ts
Comment (mariuslundgard) [reply packages/@sanity/base/src/change-indicators/helpers/getOffsetsTo.ts]: Will remove those 👍

File: packages/@sanity/desk-tool/src/panes/documentPane/types.ts
Comment (mariuslundgard) [reply packages/@sanity/desk-tool/src/panes/documentPane/types.ts:16]: Yes, will add a `@todo` there!

File: packages/@sanity/desk-tool/src/components/paneItem/helpers.tsx
Comment (mariuslundgard) [reply packages/@sanity/desk-tool/src/components/paneItem/helpers.tsx:64]: Renaming to `getPreviewStateObservable` 👍

File: packages/@sanity/desk-tool/src/components/DocTitle.tsx
Comment (mariuslundgard) [reply packages/@sanity/desk-tool/src/components/DocTitle.tsx:10]: Will look into it

File: packages/@sanity/types/parts/part.@sanity/base/other.d.ts
Comment (mariuslundgard) [reply packages/@sanity/types/parts/part.@sanity/base/other.d.ts:145]: Will look into it

File: packages/@sanity/desk-tool/src/panes/document/documentPanel/documentViews/EditForm.tsx
Comment (ricokahler) [inline packages/@sanity/desk-tool/src/panes/document/documentPanel/documentViews/EditForm.tsx:78]: it may be easier to wrap `FormBuilder` in `memo` from `'react'`. i think it's hard to manage long dependency arrays with hooks

File: packages/@sanity/desk-tool/src/panes/document/inspectDialog/InspectDialog.styles.tsx
Comment (ricokahler) [inline packages/@sanity/desk-tool/src/panes/document/inspectDialog/InspectDialog.styles.tsx:17]: just me learning: is this the preferred way to add spacing from the design system?

File: packages/@sanity/desk-tool/src/panes/document/inspectDialog/InspectDialog.styles.tsx
Comment (ricokahler) [inline packages/@sanity/desk-tool/src/panes/document/inspectDialog/InspectDialog.styles.tsx:122]: was this empty class intended?

File: packages/@sanity/desk-tool/src/panes/document/inspectDialog/InspectDialog.styles.tsx
Comment (ricokahler) [inline packages/@sanity/desk-tool/src/panes/document/inspectDialog/InspectDialog.styles.tsx:75]: 💣?

File: packages/@sanity/desk-tool/src/panes/documentList/useDocumentList.ts
Comment (ricokahler) [inline packages/@sanity/desk-tool/src/panes/documentList/useDocumentList.ts:2]: i don't think parts are being used in this file

**Review by ricokahler — APPROVED** [review]
I reviewed this one over a few days so my comments might be kind of weird! really great stuff 😎

File: packages/@sanity/desk-tool/src/utils/copyDocument.ts
Comment (mariuslundgard) [reply packages/@sanity/desk-tool/src/utils/copyDocument.ts:17]: Fixed!

File: packages/@sanity/desk-tool/src/panes/document/inspectDialog/InspectDialog.styles.tsx
Comment (mariuslundgard) [reply packages/@sanity/desk-tool/src/panes/document/inspectDialog/InspectDialog.styles.tsx:17]: No, the preferred way is to use the props like `<Box padding={4}>` when possible.

However, in this case, the elements are rendered by a 3rd party library, so we need to set padding using CSS.

Ideally, `@sanity/ui` would have CSS variables for this (e.g. `--space-4`), but that is still work-in-progress.

So yeah, this is the way to use the spacing scale in custom CSS currrently :-)

File: packages/@sanity/desk-tool/src/DeskTool.tsx
Comment (mariuslundgard) [reply packages/@sanity/desk-tool/src/DeskTool.tsx:172]: Very cool, curious to hear/see it!

File: packages/@sanity/desk-tool/src/components/pane/Pane.tsx
Comment (mariuslundgard) [reply packages/@sanity/desk-tool/src/components/pane/Pane.tsx:29]: Yes, the idea is to release it as beta components so it’s possible to use these in plugins and such in the future (or now, if you’re OK with changing APIs).

File: packages/@sanity/desk-tool/src/components/pane/Pane.tsx
Comment (mariuslundgard) [reply packages/@sanity/desk-tool/src/components/pane/Pane.tsx:129]: I think these were for debugging mainly. Will look into it, probably these can be removed.

File: packages/@sanity/desk-tool/package.json
Comment (mariuslundgard) [reply packages/@sanity/desk-tool/package.json:68]: It’s still a regular dependency :-)

**Pull request merged**
